### PR TITLE
feat: support pod propagation

### PR DESCRIPTION
**Pull request by limhawjia**
This PR supports the federation of pods using KubeAdmiral by introducing the following:

1. Add a new FTC for pods
2. Implement drop and retain logic for pod fields before syncing

Additionally:

1. The kube-scheduler is now disabled by the `dev-up.sh` script for the host cluster.
2. `dev-up.sh` now supports running 0 member clusters

### CODE REVIEW

File: config/sample/host/01-ftc.yaml [hunk 1]
```diff
@@ -692,3 +692,37 @@ spec:
     fields:
       - metadata.creationTimestamp
       - status
+---
+apiVersion: core.kubeadmiral.io/v1alpha1
```
Comment (gary-lgy) [inline config/sample/host/01-ftc.yaml:696]: Should we include it by default? This means all users will start caching pods.

File: config/sample/host/01-ftc.yaml [hunk 2]
```diff
@@ -692,3 +692,37 @@ spec:
...
+    kind: FederatedPodStatus
+    pluralName: federatedpodstatuses
+    scope: Namespaced
+    version: v1alpha1
+  controllers:
+    - - kubeadmiral.io/global-scheduler
+    - - kubeadmiral.io/overridepolicy-controller
+  statusCollection:
+    fields:
+      - status
```
Comment (gary-lgy) [inline config/sample/host/01-ftc.yaml:728]: Should we also collect spec.nodeName?

File: config/sample/host/01-ftc.yaml (hunk 1 above)
Comment (SOF3) [reply config/sample/host/01-ftc.yaml:696]: Also would it cause problems with static pods?

File: config/sample/host/01-ftc.yaml (hunk 2 above)
Comment (SOF3) [reply config/sample/host/01-ftc.yaml:729]: What happens when pod is evicted in the member?

File: config/sample/host/01-ftc.yaml (hunk 1 above)
Comment (limhawjia) [reply config/sample/host/01-ftc.yaml:696]: @gary-lgy probably not 🙂, i will move it to a separate file

**Comment by limhawjia** [comment]
> @limhawjia pls make sure to reconcile with #46, especially the interaction between pod pruning and pod propagation

The sync controller (along with other subcontrollers that will require the whole pod object) currently do not use shared informers, so it will not be affected by #46. However, I opened a new issue #64 as a reminder in the event that we do refactor these subcontrollers to use shared informers.

**Comment by limhawjia** [comment]
Pod propagation is only supported if the `BoundServiceAccountTokenVolume` is enabled (enabled by default from 1.21 onwards).

Without `BoundServiceAccountTokenVolume`, erviceaccount volumes can only be identified by checking if the volume has a secret volumeSource that references the secret containing the serviceaccount token. This would require us to send multiple requests to the apiserver or add new informers to the sync controller, adding a lot of overhead. As a result, we will not be supporting this for now.

File: pkg/controllers/sync/dispatch/retain.go [hunk 3]
```diff
@@ -251,6 +263,256 @@ func retainPersistentVolumeClaimFields(
...
+func copyUnstructuredField(srcObj, destObj *unstructured.Unstructured, fields ...string) error {
+	value, exists, err := unstructured.NestedFieldNoCopy(srcObj.Object, fields...)
+	if err != nil || !exists {
+		return nil
+	}
+
+	return unstructured.SetNestedField(destObj.Object, value, fields...)
+}
+
+// containsServiceAccountVolume checks if the given pod contains a serviceaccount volume as defined by the
```
Comment (gary-lgy) [inline pkg/controllers/sync/dispatch/retain.go:355]: Function name the comment is outdated.

File: pkg/controllers/sync/dispatch/retain.go [hunk 4]
```diff
@@ -251,6 +263,256 @@ func retainPersistentVolumeClaimFields(
...
+	if err := retainContainers(desiredInitContainers, clusterInitContainers); err != nil {
+		return err
+	}
+	if err := unstructured.SetNestedSlice(desiredObj.Object, desiredInitContainers, "spec", "initContainers"); err != nil {
+		return err
+	}
+
+	// The tolerations field is also defaulted by an admission plugin. However, we do not need to explicitly retain
+	// the default tolerations due to 2 reasons:
+	// 1. The tolerations field is mutable so any updates will not reuslt in an error
```
Comment (gary-lgy) [inline pkg/controllers/sync/dispatch/retain.go:331]:
```suggestion
	// 1. The tolerations field is mutable so any updates will not result in an error
```

File: config/sample/extra/pod-ftc.yaml
Comment (gary-lgy) [inline config/sample/extra/pod-ftc.yaml:1]: Should this file be under `config/sample/host/extra`?

File: pkg/controllers/sync/dispatch/retain.go [hunk 5]
```diff
@@ -251,6 +263,231 @@ func retainPersistentVolumeClaimFields(
...
+			copy(desiredVolumes[idx+1:], desiredVolumes[idx:])
+			desiredVolumes[idx] = volume
+
+			if err := unstructured.SetNestedSlice(desiredObj.Object, desiredVolumes, "spec", "volumes"); err != nil {
+				return err
+			}
+		}
+	}
+
+	desiredContainers, _, _ := unstructured.NestedSlice(desiredObj.Object, "spec", "containers") // this is a deepcopy
```
Comment (gary-lgy) [inline pkg/controllers/sync/dispatch/retain.go:324]: Why do we not need to handle err here?

Comment (limhawjia) [reply pkg/controllers/sync/dispatch/retain.go:324]: 1. when err is returned, desiredContainers will be `nil` which is handled by `retainContainerFields`. 
2. `unstructured.NestedSlice` returns err when the value for the given field is not a `[]interface{}`, implying that the container field in the pod is corrupted for some reason - in that case, we simply ignore the field (which brings us back to 1)

**Review by gary-lgy — APPROVED** [review]

**Pull request merged**